### ISSUE TITLE
GPUTextureFormatDescriptor.viewFormats performance hint is nonhelpful

### ISSUE THREAD

**Issue by juj** [gpuweb/gpuweb#2573]
A recent change to the spec, #2540 , added a new `viewFormats` array, in which a user must specify to let the backend know which view formats a texture might be viewed with.

It brought this kind of wording to the spec:

```
Note: Adding formats to this list may have a sizable performance impact, depending on the
user’s system. It is best to avoid adding formats unnecessarily.
```

It would be preferable to avoid language like this, since it does not help the developer to design for best performance. For example,
 - is it the just the O(n) extra number of texture formats on the view list that is the performance overhead?
 - or is it maybe that having both sRGB vs non-sRGB mixes that will be an issue?
 - or maybe float vs int? or depth vs non-depth? etc.
 - what if I don't know the actual format that user might view the texture later with? Maybe a user provided plugin decides the view later? Is adding all possible formats insane?

It is clear that `viewFormats` was added to aid developers to choose more performant paths, but now the effect is rather the opposite. The spec does not help the developer to use it performantly, unless one accumulates background knowledge beyond the spec.

The issue #168 describes a lot of that important background. A end user should not need to dig into the internals of the WebGPU issue tracker to find the important motivation behind specific options on how to use the API performantly, but the spec should carry the necessary information on its own for a developer to be able to choose the right thing.

I understand that it is appealing to want to detach the spec from all the possible hardware specifics, but I'd argue that it will just make people need to x-ray through the browser implementations and decode "how does this relate to D3D12/Vulkan/Metal" to find the right way to use the api and understand the possible tradeoffs included.

Would it be possible to have the API describe the actual details of how the feature will work on top of different backends?

**Comment by Kangz** [comment]
I don't think it is possible to have an API that accurately describes what will happen on the various backends and on the various hardware. The number of combination is huge, and various models from the same IHV have different characteristics. For example SRGB reinterpretation is expensive up to Intel Gen 9 but not afterwards. We can't encode the equivalent of [this table](https://github.com/mesa3d/mesa/blob/c572adceaa493ab2f7b6cad6d6bd100c6d93ff4d/src/intel/isl/isl_format.c#L155), for every IHV in the spec.

The note could be more clear that the expected cost is that texture compression may be disabled on some hardware if you add certain view formats, but that what causes texture compression to disappear will be very hardware dependent. If you don't know the actual format the user might use, then you're a bit out of luck, either you add all compatible formats to the compatibility list, or do a blit, or pass the actual format to the middleware and let it deal with it.

Given all that, what would be your suggestion for this? Previously we had a proposal to add just `bool srgbReinterpretation` but that doesn't cover future reinterpretation.

**Comment by kainino0x** [comment]
> The issue #168 describes a lot of that important background. A end user should not need to dig into the internals of the WebGPU issue tracker to find the important motivation behind specific options on how to use the API performantly, but the spec should carry the necessary information on its own for a developer to be able to choose the right thing.

#168 digs into two or so particular issues, but they're just guiding examples. I don't think documenting those two just because we know about them makes sense. 

The point of the note is that you shouldn't just spam the viewFormats list with tons of formats, you should be deliberate about adding only what you need. It's not intended to help you choose between entirely different codepaths that use different viewFormats for better performance, and I don't think that's possible. For a decision like that, impetus would be on the developer to benchmark different devices and architectures.

> * is it the just the O(n) extra number of texture formats on the view list that is the performance overhead?

It could be clearer though that this is not the problem.

**Comment by kvark** [comment]
Adding to other answers. If you need performance on a particular piece of hardware, *something* must know which formats are easy to cast to, and which aren't.

One choice for us is to try gathering this knowledge across all APIs and (more importantly) particular driver implementations, and encode this knowledge into the WebGPU API surface. Not only this is very hard (actually, impossible for closed source drivers, impossible for future hardware, etc), but also it doesn't actually solve it. You'd just end up with the same question just over classes of formats instead of exact formats, but essentially the problem would still be yours.

Another choice is to say we can't provide you with this insight, but we can give you tools to potentially minimize the perf loss, and to make this decision on your own. I.e. If you know the device vendor and generation, a 3rd party library can try to tell you which formats are OK to cast. I know the discussion of how much we are exposing the device ID is still highly contentious, but my understanding is that the device generation is hard to hide, generally, since you can observe the rasterization side-effects associated to them.

TL;DR: the spec note should be improved, but the API itself is already good.
